pyo3-ffi/src/dictobject.rs

```diff
@@ -96,6 +104,24 @@ pub unsafe fn PyDictViewSet_Check(op: *mut PyObject) -> c_int {
     (PyDictKeys_Check(op) != 0 || PyDictItems_Check(op) != 0) as c_int
 }
 
+#[cfg(not(Py_3_13))]
+pub unsafe fn PyDict_GetItemRef(
+    dp: *mut PyObject,
+    key: *mut PyObject,
+    result: *mut *mut PyObject,
+) -> c_int {
+    let item: *mut PyObject = PyDict_GetItemWithError(dp, key);
+    if !item.is_null() {
+        *result = _Py_NewRef(item);
+        return 1; // found
+    }
+    *result = std::ptr::null_mut();
+    if !PyErr_Occurred().is_null() {
+        return 0; // not found
+    }
+    -1
```

Comment:
This is based on the implementation in the `pythoncapi-compat` header: https://github.com/python/pythoncapi-compat/blob/ea1f7f6eac63ff401937515638252402ff33dccb/pythoncapi_compat.h#L803-L820

Reply:
This is interesting. So far we have kept these files consistent with the CPython headers as much as possible.

That said, it _is_ a desirable thing to encourage users to use this function over `PyDict_GetItem` as much as possible.

Perhaps we can add a `ffi::compat` submodule of `pyo3-ffi`, with the idea that it would always expose `PyDict_GetItemRef`? (Conditionally either adding this implementation, or re-exporting.) We could then expand this submodule as we needed in future.